notebooks/predicting_queries/entity_intent_suggestion.ipynb

```diff
@@ -0,0 +1,348 @@
+    "query_taxons = query_taxons.explode('entities')\n",
+    "query_taxons = query_taxons[['taxon','entities','tokens']]\n",
+    "query_taxons = query_taxons.dropna()\n",
+    "# Group query tokens by taxon so we have a list of tokens per taxon\n",
+    "query_taxons = query_taxons.groupby('taxon').aggregate(lambda x: list(x)).reset_index()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "title_df = pd.read_csv(PROJECT_DIR+'/data/processed/mainstreamcontent.csv')\n",
+    "title_df = title_df.dropna()\n",
+    "title_df['body_text'] = title_df['text'] + (title_df['title'] + ' ') * 5\n",
+    "title_df['body_tokens'] = title_df['body_text'].map(lambda x: tokenizer.tokenize(x.lower().translate(translator).split()))\n",
+    "title_df['entities']  = title_df['body_tokens'].map(lambda x: [entity for entity in x if entity in entity_set])\n",
+    "title_df['entity_count'] = title_df['entities'].map(Counter)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "query_taxons['entity_count'] = query_taxons['entities'].map(Counter)\n",
+    "query_boe = pd.DataFrame.from_records(query_taxons['entity_count'],index=query_taxons['taxon'])\n",
+    "query_boe = query_boe.replace(np.nan,0) \n",
+    "query_dis = query_boe.divide(query_boe.sum(1),0)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# bag of words for mainstream content\n",
+    "service_boe = pd.DataFrame.from_records(title_df['entity_count'],index=title_df['name'],columns=query_boe.columns)\n",
+    "service_boe = service_boe.replace(np.nan,0)\n",
+    "#add one smoothing\n",
+    "service_boe = service_boe +1\n",
+    "service_dis = service_boe.divide(service_boe.sum(1),0)\n",
+    "service_dis.shape"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "assert(query_dis.shape[1]==service_dis.shape[1])\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": false
+   },
+   "outputs": [],
+   "source": [
+    "# Find min KL divergence content to taxon query terms\n",
+    "# https://en.wikipedia.org/wiki/Information_projection\n",
+    "# Add one smoothing, which is p, which is q, all make quite a big difference in performance\n",
+    "# p is our normalised bag of words from each taxon query\n",
+    "# q is every bit of mainstream content normalised bag of words\n",
+    "# e.g try 'Blue badges', 'Afghanistan' etc\n",
+    "TAXON = \"Stopping or selling your business\"\n",
+    "idx = np.argsort(entropy(np.broadcast_to(query_dis.loc[TAXON].to_numpy(),(service_dis.shape[0],service_dis.shape[1])),service_dis,axis=1))[0:15]\n",
```

Comment:
I think this would benefit from splitting out and using named arguments so it is clearer.
```suggestion
    "# get entropy of a distribution for given probability values\n",
    "ent = entropy(pk=np.broadcast_to(array=query_dis.loc[TAXON].to_numpy(),\n",
                                                                "shape=service_dis.shape),\n",
                             "qk=service_dis\n",
                             "axis=1)\n",
    "idx = np.argsort(a=ent)[0:15]\n",
```